# RDM - Tools for supporting (large) data transfers

This is an experiment -- a notebook to capture practices that support the acceleration of large data transfer (10s of TBs; millions of files) and perhaps automate some or all of them.

Run this notebook locally on the machine that has to data to transfer. It's written for Mac OS X (Unix/bash). It can probably be easily modified for Linux boxes.

#### Last revised: 26 March 2018

By: Rick

## A few shell command basics

Prefix a shell command with an exclamation point ('!') to run that command from the notebook.

A few examples:

In [36]:
!python --version

Python 3.6.2 :: Anaconda custom (64-bit)


In [37]:
!pwd

/Users/rjaffe/Jupyter_notebooks/rdm_datatransfer


In [38]:
!ls

Shell_play.ipynb   dirs-gt-2000-du.sh dirs-gt-2000.sh    readme.md


Many common commands, including `pwd` and `ls`, can be run by prefixing the command with a '%' sign.

In [39]:
%cd /Users/rjaffe/Documents/RDM/RDM_Consulting/

/Users/rjaffe/Documents/RDM/RDM_Consulting


In [40]:
%cd /Users/rjaffe/Jupyter_notebooks/RDM_datatransfer

/Users/rjaffe/Jupyter_notebooks/rdm_datatransfer


These commands can be run without using the % prefix by turning on '**automagic**'.

From : https://jakevdp.github.io/PythonDataScienceHandbook/01.05-ipython-and-shell-commands.html

>Besides %cd, other available shell-like magic functions are %cat, %cp, %env, %ls, %man, %mkdir, %more, %mv, %pwd, %rm, and %rmdir, any of which can be used without the % sign if automagic is on. This makes it so that you can almost treat the IPython prompt as if it's a normal shell.

The '%automagic' command toggles the setting on and off. Run this cell several times to see. Continue with automagic on.

In [41]:
%automagic


Automagic is OFF, % prefix IS needed for line magics.


To set automagic on, follow the command with the argument '1'

In [42]:
%automagic 1


Automagic is ON, % prefix IS NOT needed for line magics.


In [43]:
pwd

'/Users/rjaffe/Jupyter_notebooks/rdm_datatransfer'

However, you still need the '!' to refer to the value returned by the command.

This works:

In [44]:
directory = !pwd

In [45]:
print(directory)

['/Users/rjaffe/Jupyter_notebooks/rdm_datatransfer']


But this throws an error:

In [46]:
directory = pwd   #This will throw an error.

NameError: name 'pwd' is not defined

Within the notebook, a variable like this has a different type than a purely Python variable.

In [47]:
type(directory)

IPython.utils.text.SList

## Sample data

To demonstrate the data profiling tools, we'll need a directory with a ton of files. I have one with Library of Congress data on my computer. Locate an appropriate set of sample data on yours.

In [48]:
cd /Users/rjaffe/documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/

/Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01


In [49]:
ls

sn89053851/ sn94051341/ sn94051342/ sn95060791/


## Some useful commands

Use the `find` command to list directories. The dot ('.') refers to your present location within the file tree. Don't forget to prefix the command with an exclamation point ("bang"). `Find` is not one of the commands covered by 'automagic'.

In [50]:
!find . -type d

.
./sn89053851
./sn89053851/1899
./sn89053851/1899/11
./sn89053851/1899/11/16
./sn89053851/1899/11/16/ed-1
./sn89053851/1899/11/16/ed-1/seq-7
./sn89053851/1899/11/16/ed-1/seq-9
./sn89053851/1899/11/16/ed-1/seq-8
./sn89053851/1899/11/16/ed-1/seq-1
./sn89053851/1899/11/16/ed-1/seq-6
./sn89053851/1899/11/16/ed-1/seq-12
./sn89053851/1899/11/16/ed-1/seq-3
./sn89053851/1899/11/16/ed-1/seq-4
./sn89053851/1899/11/16/ed-1/seq-5
./sn89053851/1899/11/16/ed-1/seq-2
./sn89053851/1899/11/16/ed-1/seq-11
./sn89053851/1899/11/16/ed-1/seq-10
./sn89053851/1899/11/30
./sn89053851/1899/11/30/ed-1
./sn89053851/1899/11/30/ed-1/seq-7
./sn89053851/1899/11/30/ed-1/seq-9
./sn89053851/1899/11/30/ed-1/seq-8
./sn89053851/1899/11/30/ed-1/seq-1
./sn89053851/1899/11/30/ed-1/seq-6
./sn89053851/1899/11/30/ed-1/seq-12
./sn89053851/1899/11/30/ed-1/seq-3
./sn89053851/1899/11/30/ed-1/seq-4
./sn89053851/1899/11/30/ed-1/seq-5
./sn89053851/1899/11/30/ed-1/seq-2
./sn89053851/1899/11/30/ed-1/seq-11
./sn89053851/1899/11/30/ed-1/s

./sn89053851/1910/07/29/ed-1
./sn89053851/1910/07/29/ed-1/seq-1
./sn89053851/1910/07/29/ed-1/seq-3
./sn89053851/1910/07/29/ed-1/seq-4
./sn89053851/1910/07/29/ed-1/seq-2
./sn89053851/1910/07/08
./sn89053851/1910/07/08/ed-1
./sn89053851/1910/07/08/ed-1/seq-1
./sn89053851/1910/07/08/ed-1/seq-3
./sn89053851/1910/07/08/ed-1/seq-4
./sn89053851/1910/07/08/ed-1/seq-2
./sn89053851/1910/07/01
./sn89053851/1910/07/01/ed-1
./sn89053851/1910/07/01/ed-1/seq-1
./sn89053851/1910/07/01/ed-1/seq-3
./sn89053851/1910/07/01/ed-1/seq-4
./sn89053851/1910/07/01/ed-1/seq-2
./sn89053851/1910/07/15
./sn89053851/1910/07/15/ed-1
./sn89053851/1910/07/15/ed-1/seq-1
./sn89053851/1910/07/15/ed-1/seq-3
./sn89053851/1910/07/15/ed-1/seq-4
./sn89053851/1910/07/15/ed-1/seq-2
./sn89053851/1910/07/22
./sn89053851/1910/07/22/ed-1
./sn89053851/1910/07/22/ed-1/seq-1
./sn89053851/1910/07/22/ed-1/seq-3
./sn89053851/1910/07/22/ed-1/seq-4
./sn89053851/1910/07/22/ed-1/seq-2
./sn89053851/1910/09
./sn89053851/1910/09/02
./sn89053851/1

./sn89053851/1905/11/09/ed-1/seq-4
./sn89053851/1905/11/09/ed-1/seq-2
./sn89053851/1905/11/30
./sn89053851/1905/11/30/ed-1
./sn89053851/1905/11/30/ed-1/seq-1
./sn89053851/1905/11/30/ed-1/seq-3
./sn89053851/1905/11/30/ed-1/seq-4
./sn89053851/1905/11/30/ed-1/seq-2
./sn89053851/1905/11/23
./sn89053851/1905/11/23/ed-1
./sn89053851/1905/11/23/ed-1/seq-1
./sn89053851/1905/11/23/ed-1/seq-3
./sn89053851/1905/11/23/ed-1/seq-4
./sn89053851/1905/11/23/ed-1/seq-2
./sn89053851/1905/10
./sn89053851/1905/10/05
./sn89053851/1905/10/05/ed-1
./sn89053851/1905/10/05/ed-1/seq-1
./sn89053851/1905/10/05/ed-1/seq-3
./sn89053851/1905/10/05/ed-1/seq-4
./sn89053851/1905/10/05/ed-1/seq-2
./sn89053851/1905/10/19
./sn89053851/1905/10/19/ed-1
./sn89053851/1905/10/19/ed-1/seq-1
./sn89053851/1905/10/19/ed-1/seq-3
./sn89053851/1905/10/19/ed-1/seq-4
./sn89053851/1905/10/19/ed-1/seq-2
./sn89053851/1905/10/26
./sn89053851/1905/10/26/ed-1
./sn89053851/1905/10/26/ed-1/seq-1
./sn89053851/1905/10/26/ed-1/seq-3
./sn89053851/1

./sn89053851/1902/09/25/ed-1/seq-1
./sn89053851/1902/09/25/ed-1/seq-3
./sn89053851/1902/09/25/ed-1/seq-4
./sn89053851/1902/09/25/ed-1/seq-2
./sn89053851/1902/08
./sn89053851/1902/08/28
./sn89053851/1902/08/28/ed-1
./sn89053851/1902/08/28/ed-1/seq-1
./sn89053851/1902/08/28/ed-1/seq-3
./sn89053851/1902/08/28/ed-1/seq-4
./sn89053851/1902/08/28/ed-1/seq-2
./sn89053851/1902/08/21
./sn89053851/1902/08/21/ed-1
./sn89053851/1902/08/21/ed-1/seq-1
./sn89053851/1902/08/21/ed-1/seq-3
./sn89053851/1902/08/21/ed-1/seq-4
./sn89053851/1902/08/21/ed-1/seq-2
./sn89053851/1902/08/07
./sn89053851/1902/08/07/ed-1
./sn89053851/1902/08/07/ed-1/seq-1
./sn89053851/1902/08/07/ed-1/seq-3
./sn89053851/1902/08/07/ed-1/seq-4
./sn89053851/1902/08/07/ed-1/seq-2
./sn89053851/1902/08/14
./sn89053851/1902/08/14/ed-1
./sn89053851/1902/08/14/ed-1/seq-1
./sn89053851/1902/08/14/ed-1/seq-3
./sn89053851/1902/08/14/ed-1/seq-4
./sn89053851/1902/08/14/ed-1/seq-2
./sn89053851/1902/01
./sn89053851/1902/01/02
./sn89053851/1902/01/0

./sn89053851/1904/03/03/ed-1/seq-3
./sn89053851/1904/03/03/ed-1/seq-4
./sn89053851/1904/03/03/ed-1/seq-2
./sn89053851/1904/03/17
./sn89053851/1904/03/17/ed-1
./sn89053851/1904/03/17/ed-1/seq-1
./sn89053851/1904/03/17/ed-1/seq-3
./sn89053851/1904/03/17/ed-1/seq-4
./sn89053851/1904/03/17/ed-1/seq-2
./sn89053851/1904/03/10
./sn89053851/1904/03/10/ed-1
./sn89053851/1904/03/10/ed-1/seq-1
./sn89053851/1904/03/10/ed-1/seq-3
./sn89053851/1904/03/10/ed-1/seq-4
./sn89053851/1904/03/10/ed-1/seq-2
./sn89053851/1904/03/31
./sn89053851/1904/03/31/ed-1
./sn89053851/1904/03/31/ed-1/seq-1
./sn89053851/1904/03/31/ed-1/seq-3
./sn89053851/1904/03/31/ed-1/seq-4
./sn89053851/1904/03/31/ed-1/seq-2
./sn89053851/1904/03/24
./sn89053851/1904/03/24/ed-1
./sn89053851/1904/03/24/ed-1/seq-1
./sn89053851/1904/03/24/ed-1/seq-3
./sn89053851/1904/03/24/ed-1/seq-4
./sn89053851/1904/03/24/ed-1/seq-2
./sn89053851/1904/04
./sn89053851/1904/04/28
./sn89053851/1904/04/28/ed-1
./sn89053851/1904/04/28/ed-1/seq-1
./sn89053851/1

./sn89053851/1901/12/26/ed-1/seq-1
./sn89053851/1901/12/26/ed-1/seq-6
./sn89053851/1901/12/26/ed-1/seq-3
./sn89053851/1901/12/26/ed-1/seq-4
./sn89053851/1901/12/26/ed-1/seq-5
./sn89053851/1901/12/26/ed-1/seq-2
./sn89053851/1901/12/12
./sn89053851/1901/12/12/ed-1
./sn89053851/1901/12/12/ed-1/seq-1
./sn89053851/1901/12/12/ed-1/seq-3
./sn89053851/1901/12/12/ed-1/seq-4
./sn89053851/1901/12/12/ed-1/seq-2
./sn89053851/1906
./sn89053851/1906/03
./sn89053851/1906/03/29
./sn89053851/1906/03/29/ed-1
./sn89053851/1906/03/29/ed-1/seq-1
./sn89053851/1906/03/29/ed-1/seq-3
./sn89053851/1906/03/29/ed-1/seq-4
./sn89053851/1906/03/29/ed-1/seq-2
./sn89053851/1906/03/08
./sn89053851/1906/03/08/ed-1
./sn89053851/1906/03/08/ed-1/seq-1
./sn89053851/1906/03/08/ed-1/seq-3
./sn89053851/1906/03/08/ed-1/seq-4
./sn89053851/1906/03/08/ed-1/seq-2
./sn89053851/1906/03/01
./sn89053851/1906/03/01/ed-1
./sn89053851/1906/03/01/ed-1/seq-1
./sn89053851/1906/03/01/ed-1/seq-3
./sn89053851/1906/03/01/ed-1/seq-4
./sn89053851/1

./sn89053851/1908/11
./sn89053851/1908/11/05
./sn89053851/1908/11/05/ed-1
./sn89053851/1908/11/05/ed-1/seq-1
./sn89053851/1908/11/05/ed-1/seq-3
./sn89053851/1908/11/05/ed-1/seq-4
./sn89053851/1908/11/05/ed-1/seq-2
./sn89053851/1908/11/19
./sn89053851/1908/11/19/ed-1
./sn89053851/1908/11/19/ed-1/seq-1
./sn89053851/1908/11/19/ed-1/seq-3
./sn89053851/1908/11/19/ed-1/seq-4
./sn89053851/1908/11/19/ed-1/seq-2
./sn89053851/1908/11/26
./sn89053851/1908/11/26/ed-1
./sn89053851/1908/11/26/ed-1/seq-1
./sn89053851/1908/11/26/ed-1/seq-3
./sn89053851/1908/11/26/ed-1/seq-4
./sn89053851/1908/11/26/ed-1/seq-2
./sn89053851/1908/11/12
./sn89053851/1908/11/12/ed-1
./sn89053851/1908/11/12/ed-1/seq-1
./sn89053851/1908/11/12/ed-1/seq-3
./sn89053851/1908/11/12/ed-1/seq-4
./sn89053851/1908/11/12/ed-1/seq-2
./sn89053851/1908/10
./sn89053851/1908/10/29
./sn89053851/1908/10/29/ed-1
./sn89053851/1908/10/29/ed-1/seq-1
./sn89053851/1908/10/29/ed-1/seq-3
./sn89053851/1908/10/29/ed-1/seq-4
./sn89053851/1908/10/29/ed-1

./sn89053851/1909/08/05/ed-1
./sn89053851/1909/08/05/ed-1/seq-1
./sn89053851/1909/08/05/ed-1/seq-3
./sn89053851/1909/08/05/ed-1/seq-4
./sn89053851/1909/08/05/ed-1/seq-2
./sn89053851/1909/08/19
./sn89053851/1909/08/19/ed-1
./sn89053851/1909/08/19/ed-1/seq-1
./sn89053851/1909/08/19/ed-1/seq-3
./sn89053851/1909/08/19/ed-1/seq-4
./sn89053851/1909/08/19/ed-1/seq-2
./sn89053851/1909/08/26
./sn89053851/1909/08/26/ed-1
./sn89053851/1909/08/26/ed-1/seq-1
./sn89053851/1909/08/26/ed-1/seq-3
./sn89053851/1909/08/26/ed-1/seq-4
./sn89053851/1909/08/26/ed-1/seq-2
./sn89053851/1909/08/12
./sn89053851/1909/08/12/ed-1
./sn89053851/1909/08/12/ed-1/seq-1
./sn89053851/1909/08/12/ed-1/seq-3
./sn89053851/1909/08/12/ed-1/seq-4
./sn89053851/1909/08/12/ed-1/seq-2
./sn89053851/1909/01
./sn89053851/1909/01/28
./sn89053851/1909/01/28/ed-1
./sn89053851/1909/01/28/ed-1/seq-1
./sn89053851/1909/01/28/ed-1/seq-3
./sn89053851/1909/01/28/ed-1/seq-4
./sn89053851/1909/01/28/ed-1/seq-2
./sn89053851/1909/01/21
./sn89053851/1

./sn89053851/1907/12/05/ed-1
./sn89053851/1907/12/05/ed-1/seq-1
./sn89053851/1907/12/05/ed-1/seq-3
./sn89053851/1907/12/05/ed-1/seq-4
./sn89053851/1907/12/05/ed-1/seq-2
./sn89053851/1907/12/19
./sn89053851/1907/12/19/ed-1
./sn89053851/1907/12/19/ed-1/seq-1
./sn89053851/1907/12/19/ed-1/seq-3
./sn89053851/1907/12/19/ed-1/seq-4
./sn89053851/1907/12/19/ed-1/seq-2
./sn89053851/1907/12/26
./sn89053851/1907/12/26/ed-1
./sn89053851/1907/12/26/ed-1/seq-1
./sn89053851/1907/12/26/ed-1/seq-6
./sn89053851/1907/12/26/ed-1/seq-3
./sn89053851/1907/12/26/ed-1/seq-4
./sn89053851/1907/12/26/ed-1/seq-5
./sn89053851/1907/12/26/ed-1/seq-2
./sn89053851/1907/12/12
./sn89053851/1907/12/12/ed-1
./sn89053851/1907/12/12/ed-1/seq-1
./sn89053851/1907/12/12/ed-1/seq-3
./sn89053851/1907/12/12/ed-1/seq-4
./sn89053851/1907/12/12/ed-1/seq-2
./sn89053851/1900
./sn89053851/1900/03
./sn89053851/1900/03/01
./sn89053851/1900/03/01/ed-1
./sn89053851/1900/03/01/ed-1/seq-7
./sn89053851/1900/03/01/ed-1/seq-9
./sn89053851/1900/03

./sn94051342/1910/02/22/ed-1/seq-3
./sn94051342/1910/02/22/ed-1/seq-4
./sn94051342/1910/02/22/ed-1/seq-2
./sn94051342/1910/11
./sn94051342/1910/11/29
./sn94051342/1910/11/29/ed-1
./sn94051342/1910/11/29/ed-1/seq-1
./sn94051342/1910/11/29/ed-1/seq-3
./sn94051342/1910/11/29/ed-1/seq-4
./sn94051342/1910/11/29/ed-1/seq-2
./sn94051342/1910/11/08
./sn94051342/1910/11/08/ed-1
./sn94051342/1910/11/08/ed-1/seq-1
./sn94051342/1910/11/08/ed-1/seq-3
./sn94051342/1910/11/08/ed-1/seq-4
./sn94051342/1910/11/08/ed-1/seq-2
./sn94051342/1910/11/01
./sn94051342/1910/11/01/ed-1
./sn94051342/1910/11/01/ed-1/seq-1
./sn94051342/1910/11/01/ed-1/seq-3
./sn94051342/1910/11/01/ed-1/seq-4
./sn94051342/1910/11/01/ed-1/seq-2
./sn94051342/1910/11/15
./sn94051342/1910/11/15/ed-1
./sn94051342/1910/11/15/ed-1/seq-1
./sn94051342/1910/11/15/ed-1/seq-3
./sn94051342/1910/11/15/ed-1/seq-4
./sn94051342/1910/11/15/ed-1/seq-2
./sn94051342/1910/11/22
./sn94051342/1910/11/22/ed-1
./sn94051342/1910/11/22/ed-1/seq-1
./sn94051342/1

./sn94051342/1911/11/14/ed-1/seq-4
./sn94051342/1911/11/14/ed-1/seq-2
./sn94051342/1911/10
./sn94051342/1911/10/03
./sn94051342/1911/10/03/ed-1
./sn94051342/1911/10/03/ed-1/seq-1
./sn94051342/1911/10/03/ed-1/seq-3
./sn94051342/1911/10/03/ed-1/seq-4
./sn94051342/1911/10/03/ed-1/seq-2
./sn94051342/1911/10/17
./sn94051342/1911/10/17/ed-1
./sn94051342/1911/10/17/ed-1/seq-1
./sn94051342/1911/10/17/ed-1/seq-3
./sn94051342/1911/10/17/ed-1/seq-4
./sn94051342/1911/10/17/ed-1/seq-2
./sn94051342/1911/10/10
./sn94051342/1911/10/10/ed-1
./sn94051342/1911/10/10/ed-1/seq-1
./sn94051342/1911/10/10/ed-1/seq-3
./sn94051342/1911/10/10/ed-1/seq-4
./sn94051342/1911/10/10/ed-1/seq-2
./sn94051342/1911/10/31
./sn94051342/1911/10/31/ed-1
./sn94051342/1911/10/31/ed-1/seq-1
./sn94051342/1911/10/31/ed-1/seq-3
./sn94051342/1911/10/31/ed-1/seq-4
./sn94051342/1911/10/31/ed-1/seq-2
./sn94051342/1911/10/24
./sn94051342/1911/10/24/ed-1
./sn94051342/1911/10/24/ed-1/seq-1
./sn94051342/1911/10/24/ed-1/seq-3
./sn94051342/1

./sn94051342/1912/02/06
./sn94051342/1912/02/06/ed-1
./sn94051342/1912/02/06/ed-1/seq-1
./sn94051342/1912/02/06/ed-1/seq-3
./sn94051342/1912/02/06/ed-1/seq-4
./sn94051342/1912/02/06/ed-1/seq-2
./sn94051342/1912/02/13
./sn94051342/1912/02/13/ed-1
./sn94051342/1912/02/13/ed-1/seq-1
./sn94051342/1912/02/13/ed-1/seq-3
./sn94051342/1912/02/13/ed-1/seq-4
./sn94051342/1912/02/13/ed-1/seq-2
./sn94051342/1912/11
./sn94051342/1912/11/05
./sn94051342/1912/11/05/ed-1
./sn94051342/1912/11/05/ed-1/seq-1
./sn94051342/1912/11/05/ed-1/seq-3
./sn94051342/1912/11/05/ed-1/seq-4
./sn94051342/1912/11/05/ed-1/seq-2
./sn94051342/1912/11/19
./sn94051342/1912/11/19/ed-1
./sn94051342/1912/11/19/ed-1/seq-1
./sn94051342/1912/11/19/ed-1/seq-3
./sn94051342/1912/11/19/ed-1/seq-4
./sn94051342/1912/11/19/ed-1/seq-2
./sn94051342/1912/11/26
./sn94051342/1912/11/26/ed-1
./sn94051342/1912/11/26/ed-1/seq-1
./sn94051342/1912/11/26/ed-1/seq-3
./sn94051342/1912/11/26/ed-1/seq-4
./sn94051342/1912/11/26/ed-1/seq-2
./sn94051342/1

./sn94051342/1901/07/20
./sn94051342/1901/07/20/ed-1
./sn94051342/1901/07/20/ed-1/seq-1
./sn94051342/1901/07/20/ed-1/seq-3
./sn94051342/1901/07/20/ed-1/seq-4
./sn94051342/1901/07/20/ed-1/seq-2
./sn94051342/1901/07/27
./sn94051342/1901/07/27/ed-1
./sn94051342/1901/07/27/ed-1/seq-1
./sn94051342/1901/07/27/ed-1/seq-3
./sn94051342/1901/07/27/ed-1/seq-4
./sn94051342/1901/07/27/ed-1/seq-2
./sn94051342/1901/07/06
./sn94051342/1901/07/06/ed-1
./sn94051342/1901/07/06/ed-1/seq-1
./sn94051342/1901/07/06/ed-1/seq-3
./sn94051342/1901/07/06/ed-1/seq-4
./sn94051342/1901/07/06/ed-1/seq-2
./sn94051342/1901/07/13
./sn94051342/1901/07/13/ed-1
./sn94051342/1901/07/13/ed-1/seq-1
./sn94051342/1901/07/13/ed-1/seq-3
./sn94051342/1901/07/13/ed-1/seq-4
./sn94051342/1901/07/13/ed-1/seq-2
./sn94051342/1901/09
./sn94051342/1901/09/28
./sn94051342/1901/09/28/ed-1
./sn94051342/1901/09/28/ed-1/seq-1
./sn94051342/1901/09/28/ed-1/seq-3
./sn94051342/1901/09/28/ed-1/seq-4
./sn94051342/1901/09/28/ed-1/seq-2
./sn94051342/1

./sn94051342/1908/09/22/ed-1/seq-2
./sn94051342/1908/08
./sn94051342/1908/08/04
./sn94051342/1908/08/04/ed-1
./sn94051342/1908/08/04/ed-1/seq-1
./sn94051342/1908/08/04/ed-1/seq-6
./sn94051342/1908/08/04/ed-1/seq-3
./sn94051342/1908/08/04/ed-1/seq-4
./sn94051342/1908/08/04/ed-1/seq-5
./sn94051342/1908/08/04/ed-1/seq-2
./sn94051342/1908/08/18
./sn94051342/1908/08/18/ed-1
./sn94051342/1908/08/18/ed-1/seq-1
./sn94051342/1908/08/18/ed-1/seq-3
./sn94051342/1908/08/18/ed-1/seq-4
./sn94051342/1908/08/18/ed-1/seq-2
./sn94051342/1908/08/11
./sn94051342/1908/08/11/ed-1
./sn94051342/1908/08/11/ed-1/seq-1
./sn94051342/1908/08/11/ed-1/seq-6
./sn94051342/1908/08/11/ed-1/seq-3
./sn94051342/1908/08/11/ed-1/seq-4
./sn94051342/1908/08/11/ed-1/seq-5
./sn94051342/1908/08/11/ed-1/seq-2
./sn94051342/1908/08/25
./sn94051342/1908/08/25/ed-1
./sn94051342/1908/08/25/ed-1/seq-1
./sn94051342/1908/08/25/ed-1/seq-3
./sn94051342/1908/08/25/ed-1/seq-4
./sn94051342/1908/08/25/ed-1/seq-2
./sn94051342/1908/01
./sn9405134

./sn94051342/1907/03/05
./sn94051342/1907/03/05/ed-1
./sn94051342/1907/03/05/ed-1/seq-1
./sn94051342/1907/03/05/ed-1/seq-6
./sn94051342/1907/03/05/ed-1/seq-3
./sn94051342/1907/03/05/ed-1/seq-4
./sn94051342/1907/03/05/ed-1/seq-5
./sn94051342/1907/03/05/ed-1/seq-2
./sn94051342/1907/03/19
./sn94051342/1907/03/19/ed-1
./sn94051342/1907/03/19/ed-1/seq-1
./sn94051342/1907/03/19/ed-1/seq-6
./sn94051342/1907/03/19/ed-1/seq-3
./sn94051342/1907/03/19/ed-1/seq-4
./sn94051342/1907/03/19/ed-1/seq-5
./sn94051342/1907/03/19/ed-1/seq-2
./sn94051342/1907/03/26
./sn94051342/1907/03/26/ed-1
./sn94051342/1907/03/26/ed-1/seq-1
./sn94051342/1907/03/26/ed-1/seq-6
./sn94051342/1907/03/26/ed-1/seq-3
./sn94051342/1907/03/26/ed-1/seq-4
./sn94051342/1907/03/26/ed-1/seq-5
./sn94051342/1907/03/26/ed-1/seq-2
./sn94051342/1907/03/12
./sn94051342/1907/03/12/ed-1
./sn94051342/1907/03/12/ed-1/seq-1
./sn94051342/1907/03/12/ed-1/seq-6
./sn94051342/1907/03/12/ed-1/seq-3
./sn94051342/1907/03/12/ed-1/seq-4
./sn94051342/1907/

./sn94051342/1900/05
./sn94051342/1900/05/19
./sn94051342/1900/05/19/ed-1
./sn94051342/1900/05/19/ed-1/seq-1
./sn94051342/1900/05/19/ed-1/seq-3
./sn94051342/1900/05/19/ed-1/seq-4
./sn94051342/1900/05/19/ed-1/seq-2
./sn94051342/1900/05/26
./sn94051342/1900/05/26/ed-1
./sn94051342/1900/05/26/ed-1/seq-1
./sn94051342/1900/05/26/ed-1/seq-3
./sn94051342/1900/05/26/ed-1/seq-4
./sn94051342/1900/05/26/ed-1/seq-2
./sn94051342/1900/05/12
./sn94051342/1900/05/12/ed-1
./sn94051342/1900/05/12/ed-1/seq-1
./sn94051342/1900/05/12/ed-1/seq-3
./sn94051342/1900/05/12/ed-1/seq-4
./sn94051342/1900/05/12/ed-1/seq-2
./sn94051342/1900/11
./sn94051342/1900/11/03
./sn94051342/1900/11/03/ed-1
./sn94051342/1900/11/03/ed-1/seq-1
./sn94051342/1900/11/03/ed-1/seq-3
./sn94051342/1900/11/03/ed-1/seq-4
./sn94051342/1900/11/03/ed-1/seq-2
./sn94051342/1900/11/17
./sn94051342/1900/11/17/ed-1
./sn94051342/1900/11/17/ed-1/seq-1
./sn94051342/1900/11/17/ed-1/seq-3
./sn94051342/1900/11/17/ed-1/seq-4
./sn94051342/1900/11/17/ed-1

./sn94051341/1899/10/21/ed-1/seq-7
./sn94051341/1899/10/21/ed-1/seq-8
./sn94051341/1899/10/21/ed-1/seq-1
./sn94051341/1899/10/21/ed-1/seq-6
./sn94051341/1899/10/21/ed-1/seq-3
./sn94051341/1899/10/21/ed-1/seq-4
./sn94051341/1899/10/21/ed-1/seq-5
./sn94051341/1899/10/21/ed-1/seq-2
./sn94051341/1899/10/07
./sn94051341/1899/10/07/ed-1
./sn94051341/1899/10/07/ed-1/seq-7
./sn94051341/1899/10/07/ed-1/seq-8
./sn94051341/1899/10/07/ed-1/seq-1
./sn94051341/1899/10/07/ed-1/seq-6
./sn94051341/1899/10/07/ed-1/seq-3
./sn94051341/1899/10/07/ed-1/seq-4
./sn94051341/1899/10/07/ed-1/seq-5
./sn94051341/1899/10/07/ed-1/seq-2
./sn94051341/1899/10/14
./sn94051341/1899/10/14/ed-1
./sn94051341/1899/10/14/ed-1/seq-7
./sn94051341/1899/10/14/ed-1/seq-8
./sn94051341/1899/10/14/ed-1/seq-1
./sn94051341/1899/10/14/ed-1/seq-6
./sn94051341/1899/10/14/ed-1/seq-3
./sn94051341/1899/10/14/ed-1/seq-4
./sn94051341/1899/10/14/ed-1/seq-5
./sn94051341/1899/10/14/ed-1/seq-2
./sn94051341/1899/07
./sn94051341/1899/07/29
./sn94051

./sn94051341/1897/10/30/ed-1/seq-5
./sn94051341/1897/10/30/ed-1/seq-2
./sn94051341/1897/10/23
./sn94051341/1897/10/23/ed-1
./sn94051341/1897/10/23/ed-1/seq-7
./sn94051341/1897/10/23/ed-1/seq-8
./sn94051341/1897/10/23/ed-1/seq-1
./sn94051341/1897/10/23/ed-1/seq-6
./sn94051341/1897/10/23/ed-1/seq-3
./sn94051341/1897/10/23/ed-1/seq-4
./sn94051341/1897/10/23/ed-1/seq-5
./sn94051341/1897/10/23/ed-1/seq-2
./sn94051341/1897/07
./sn94051341/1897/07/03
./sn94051341/1897/07/03/ed-1
./sn94051341/1897/07/03/ed-1/seq-1
./sn94051341/1897/07/03/ed-1/seq-3
./sn94051341/1897/07/03/ed-1/seq-4
./sn94051341/1897/07/03/ed-1/seq-2
./sn94051341/1897/07/17
./sn94051341/1897/07/17/ed-1
./sn94051341/1897/07/17/ed-1/seq-1
./sn94051341/1897/07/17/ed-1/seq-3
./sn94051341/1897/07/17/ed-1/seq-4
./sn94051341/1897/07/17/ed-1/seq-2
./sn94051341/1897/07/10
./sn94051341/1897/07/10/ed-1
./sn94051341/1897/07/10/ed-1/seq-1
./sn94051341/1897/07/10/ed-1/seq-3
./sn94051341/1897/07/10/ed-1/seq-4
./sn94051341/1897/07/10/ed-1/seq

./sn94051341/1896/11/28/ed-1/seq-1
./sn94051341/1896/11/28/ed-1/seq-3
./sn94051341/1896/11/28/ed-1/seq-4
./sn94051341/1896/11/28/ed-1/seq-2
./sn94051341/1896/11/21
./sn94051341/1896/11/21/ed-1
./sn94051341/1896/11/21/ed-1/seq-1
./sn94051341/1896/11/21/ed-1/seq-3
./sn94051341/1896/11/21/ed-1/seq-4
./sn94051341/1896/11/21/ed-1/seq-2
./sn94051341/1896/11/07
./sn94051341/1896/11/07/ed-1
./sn94051341/1896/11/07/ed-1/seq-1
./sn94051341/1896/11/07/ed-1/seq-3
./sn94051341/1896/11/07/ed-1/seq-4
./sn94051341/1896/11/07/ed-1/seq-2
./sn94051341/1896/11/14
./sn94051341/1896/11/14/ed-1
./sn94051341/1896/11/14/ed-1/seq-1
./sn94051341/1896/11/14/ed-1/seq-3
./sn94051341/1896/11/14/ed-1/seq-4
./sn94051341/1896/11/14/ed-1/seq-2
./sn94051341/1896/10
./sn94051341/1896/10/03
./sn94051341/1896/10/03/ed-1
./sn94051341/1896/10/03/ed-1/seq-1
./sn94051341/1896/10/03/ed-1/seq-3
./sn94051341/1896/10/03/ed-1/seq-4
./sn94051341/1896/10/03/ed-1/seq-2
./sn94051341/1896/10/17
./sn94051341/1896/10/17/ed-1
./sn94051341/1

./sn94051341/1898/10/08/ed-1/seq-2
./sn94051341/1898/10/01
./sn94051341/1898/10/01/ed-1
./sn94051341/1898/10/01/ed-1/seq-7
./sn94051341/1898/10/01/ed-1/seq-8
./sn94051341/1898/10/01/ed-1/seq-1
./sn94051341/1898/10/01/ed-1/seq-6
./sn94051341/1898/10/01/ed-1/seq-3
./sn94051341/1898/10/01/ed-1/seq-4
./sn94051341/1898/10/01/ed-1/seq-5
./sn94051341/1898/10/01/ed-1/seq-2
./sn94051341/1898/10/15
./sn94051341/1898/10/15/ed-1
./sn94051341/1898/10/15/ed-1/seq-7
./sn94051341/1898/10/15/ed-1/seq-8
./sn94051341/1898/10/15/ed-1/seq-1
./sn94051341/1898/10/15/ed-1/seq-6
./sn94051341/1898/10/15/ed-1/seq-3
./sn94051341/1898/10/15/ed-1/seq-4
./sn94051341/1898/10/15/ed-1/seq-5
./sn94051341/1898/10/15/ed-1/seq-2
./sn94051341/1898/10/22
./sn94051341/1898/10/22/ed-1
./sn94051341/1898/10/22/ed-1/seq-7
./sn94051341/1898/10/22/ed-1/seq-8
./sn94051341/1898/10/22/ed-1/seq-1
./sn94051341/1898/10/22/ed-1/seq-6
./sn94051341/1898/10/22/ed-1/seq-3
./sn94051341/1898/10/22/ed-1/seq-4
./sn94051341/1898/10/22/ed-1/seq-5
.

./sn94051341/1900/03/03
./sn94051341/1900/03/03/ed-1
./sn94051341/1900/03/03/ed-1/seq-7
./sn94051341/1900/03/03/ed-1/seq-8
./sn94051341/1900/03/03/ed-1/seq-1
./sn94051341/1900/03/03/ed-1/seq-6
./sn94051341/1900/03/03/ed-1/seq-3
./sn94051341/1900/03/03/ed-1/seq-4
./sn94051341/1900/03/03/ed-1/seq-5
./sn94051341/1900/03/03/ed-1/seq-2
./sn94051341/1900/03/17
./sn94051341/1900/03/17/ed-1
./sn94051341/1900/03/17/ed-1/seq-7
./sn94051341/1900/03/17/ed-1/seq-8
./sn94051341/1900/03/17/ed-1/seq-1
./sn94051341/1900/03/17/ed-1/seq-6
./sn94051341/1900/03/17/ed-1/seq-3
./sn94051341/1900/03/17/ed-1/seq-4
./sn94051341/1900/03/17/ed-1/seq-5
./sn94051341/1900/03/17/ed-1/seq-2
./sn94051341/1900/03/10
./sn94051341/1900/03/10/ed-1
./sn94051341/1900/03/10/ed-1/seq-7
./sn94051341/1900/03/10/ed-1/seq-8
./sn94051341/1900/03/10/ed-1/seq-1
./sn94051341/1900/03/10/ed-1/seq-6
./sn94051341/1900/03/10/ed-1/seq-3
./sn94051341/1900/03/10/ed-1/seq-4
./sn94051341/1900/03/10/ed-1/seq-5
./sn94051341/1900/03/10/ed-1/seq-2
.

./sn95060791/1919/05/30/ed-1
./sn95060791/1919/05/30/ed-1/seq-1
./sn95060791/1919/05/30/ed-1/seq-3
./sn95060791/1919/05/30/ed-1/seq-4
./sn95060791/1919/05/30/ed-1/seq-2
./sn95060791/1919/05/23
./sn95060791/1919/05/23/ed-1
./sn95060791/1919/05/23/ed-1/seq-1
./sn95060791/1919/05/23/ed-1/seq-3
./sn95060791/1919/05/23/ed-1/seq-4
./sn95060791/1919/05/23/ed-1/seq-2
./sn95060791/1919/02
./sn95060791/1919/02/28
./sn95060791/1919/02/28/ed-1
./sn95060791/1919/02/28/ed-1/seq-1
./sn95060791/1919/02/28/ed-1/seq-3
./sn95060791/1919/02/28/ed-1/seq-4
./sn95060791/1919/02/28/ed-1/seq-2
./sn95060791/1919/02/21
./sn95060791/1919/02/21/ed-1
./sn95060791/1919/02/21/ed-1/seq-1
./sn95060791/1919/02/21/ed-1/seq-3
./sn95060791/1919/02/21/ed-1/seq-4
./sn95060791/1919/02/21/ed-1/seq-2
./sn95060791/1919/02/07
./sn95060791/1919/02/07/ed-1
./sn95060791/1919/02/07/ed-1/seq-1
./sn95060791/1919/02/07/ed-1/seq-3
./sn95060791/1919/02/07/ed-1/seq-4
./sn95060791/1919/02/07/ed-1/seq-2
./sn95060791/1919/02/14
./sn95060791/1

./sn95060791/1921/02/11
./sn95060791/1921/02/11/ed-1
./sn95060791/1921/02/11/ed-1/seq-7
./sn95060791/1921/02/11/ed-1/seq-8
./sn95060791/1921/02/11/ed-1/seq-1
./sn95060791/1921/02/11/ed-1/seq-6
./sn95060791/1921/02/11/ed-1/seq-3
./sn95060791/1921/02/11/ed-1/seq-4
./sn95060791/1921/02/11/ed-1/seq-5
./sn95060791/1921/02/11/ed-1/seq-2
./sn95060791/1921/02/25
./sn95060791/1921/02/25/ed-1
./sn95060791/1921/02/25/ed-1/seq-7
./sn95060791/1921/02/25/ed-1/seq-8
./sn95060791/1921/02/25/ed-1/seq-1
./sn95060791/1921/02/25/ed-1/seq-6
./sn95060791/1921/02/25/ed-1/seq-3
./sn95060791/1921/02/25/ed-1/seq-4
./sn95060791/1921/02/25/ed-1/seq-5
./sn95060791/1921/02/25/ed-1/seq-2
./sn95060791/1921/11
./sn95060791/1921/11/04
./sn95060791/1921/11/04/ed-1
./sn95060791/1921/11/04/ed-1/seq-1
./sn95060791/1921/11/04/ed-1/seq-3
./sn95060791/1921/11/04/ed-1/seq-4
./sn95060791/1921/11/04/ed-1/seq-2
./sn95060791/1921/11/18
./sn95060791/1921/11/18/ed-1
./sn95060791/1921/11/18/ed-1/seq-1
./sn95060791/1921/11/18/ed-1/seq

./sn95060791/1921/06/10/ed-1/seq-4
./sn95060791/1921/06/10/ed-1/seq-5
./sn95060791/1921/06/10/ed-1/seq-2
./sn95060791/1921/06/24
./sn95060791/1921/06/24/ed-1
./sn95060791/1921/06/24/ed-1/seq-7
./sn95060791/1921/06/24/ed-1/seq-8
./sn95060791/1921/06/24/ed-1/seq-1
./sn95060791/1921/06/24/ed-1/seq-6
./sn95060791/1921/06/24/ed-1/seq-3
./sn95060791/1921/06/24/ed-1/seq-4
./sn95060791/1921/06/24/ed-1/seq-5
./sn95060791/1921/06/24/ed-1/seq-2
./sn95060791/1921/12
./sn95060791/1921/12/02
./sn95060791/1921/12/02/ed-1
./sn95060791/1921/12/02/ed-1/seq-1
./sn95060791/1921/12/02/ed-1/seq-3
./sn95060791/1921/12/02/ed-1/seq-4
./sn95060791/1921/12/02/ed-1/seq-2
./sn95060791/1921/12/16
./sn95060791/1921/12/16/ed-1
./sn95060791/1921/12/16/ed-1/seq-1
./sn95060791/1921/12/16/ed-1/seq-3
./sn95060791/1921/12/16/ed-1/seq-4
./sn95060791/1921/12/16/ed-1/seq-2
./sn95060791/1921/12/09
./sn95060791/1921/12/09/ed-1
./sn95060791/1921/12/09/ed-1/seq-1
./sn95060791/1921/12/09/ed-1/seq-3
./sn95060791/1921/12/09/ed-1/seq

./sn95060791/1910/04/01/ed-1/seq-7
./sn95060791/1910/04/01/ed-1/seq-8
./sn95060791/1910/04/01/ed-1/seq-1
./sn95060791/1910/04/01/ed-1/seq-6
./sn95060791/1910/04/01/ed-1/seq-3
./sn95060791/1910/04/01/ed-1/seq-4
./sn95060791/1910/04/01/ed-1/seq-5
./sn95060791/1910/04/01/ed-1/seq-2
./sn95060791/1910/04/15
./sn95060791/1910/04/15/ed-1
./sn95060791/1910/04/15/ed-1/seq-7
./sn95060791/1910/04/15/ed-1/seq-8
./sn95060791/1910/04/15/ed-1/seq-1
./sn95060791/1910/04/15/ed-1/seq-6
./sn95060791/1910/04/15/ed-1/seq-3
./sn95060791/1910/04/15/ed-1/seq-4
./sn95060791/1910/04/15/ed-1/seq-5
./sn95060791/1910/04/15/ed-1/seq-2
./sn95060791/1910/04/22
./sn95060791/1910/04/22/ed-1
./sn95060791/1910/04/22/ed-1/seq-7
./sn95060791/1910/04/22/ed-1/seq-8
./sn95060791/1910/04/22/ed-1/seq-1
./sn95060791/1910/04/22/ed-1/seq-6
./sn95060791/1910/04/22/ed-1/seq-3
./sn95060791/1910/04/22/ed-1/seq-4
./sn95060791/1910/04/22/ed-1/seq-5
./sn95060791/1910/04/22/ed-1/seq-2
./sn95060791/1910/05
./sn95060791/1910/05/20
./sn95060

./sn95060791/1911/03/24/ed-1/seq-7
./sn95060791/1911/03/24/ed-1/seq-8
./sn95060791/1911/03/24/ed-1/seq-1
./sn95060791/1911/03/24/ed-1/seq-6
./sn95060791/1911/03/24/ed-1/seq-3
./sn95060791/1911/03/24/ed-1/seq-4
./sn95060791/1911/03/24/ed-1/seq-5
./sn95060791/1911/03/24/ed-1/seq-2
./sn95060791/1911/04
./sn95060791/1911/04/28
./sn95060791/1911/04/28/ed-1
./sn95060791/1911/04/28/ed-1/seq-7
./sn95060791/1911/04/28/ed-1/seq-8
./sn95060791/1911/04/28/ed-1/seq-1
./sn95060791/1911/04/28/ed-1/seq-6
./sn95060791/1911/04/28/ed-1/seq-3
./sn95060791/1911/04/28/ed-1/seq-4
./sn95060791/1911/04/28/ed-1/seq-5
./sn95060791/1911/04/28/ed-1/seq-2
./sn95060791/1911/04/21
./sn95060791/1911/04/21/ed-1
./sn95060791/1911/04/21/ed-1/seq-7
./sn95060791/1911/04/21/ed-1/seq-8
./sn95060791/1911/04/21/ed-1/seq-1
./sn95060791/1911/04/21/ed-1/seq-6
./sn95060791/1911/04/21/ed-1/seq-3
./sn95060791/1911/04/21/ed-1/seq-4
./sn95060791/1911/04/21/ed-1/seq-5
./sn95060791/1911/04/21/ed-1/seq-2
./sn95060791/1911/04/07
./sn95060

./sn95060791/1911/01/20/ed-1
./sn95060791/1911/01/20/ed-1/seq-7
./sn95060791/1911/01/20/ed-1/seq-8
./sn95060791/1911/01/20/ed-1/seq-1
./sn95060791/1911/01/20/ed-1/seq-6
./sn95060791/1911/01/20/ed-1/seq-3
./sn95060791/1911/01/20/ed-1/seq-4
./sn95060791/1911/01/20/ed-1/seq-5
./sn95060791/1911/01/20/ed-1/seq-2
./sn95060791/1911/01/27
./sn95060791/1911/01/27/ed-1
./sn95060791/1911/01/27/ed-1/seq-7
./sn95060791/1911/01/27/ed-1/seq-8
./sn95060791/1911/01/27/ed-1/seq-1
./sn95060791/1911/01/27/ed-1/seq-6
./sn95060791/1911/01/27/ed-1/seq-3
./sn95060791/1911/01/27/ed-1/seq-4
./sn95060791/1911/01/27/ed-1/seq-5
./sn95060791/1911/01/27/ed-1/seq-2
./sn95060791/1911/01/06
./sn95060791/1911/01/06/ed-1
./sn95060791/1911/01/06/ed-1/seq-7
./sn95060791/1911/01/06/ed-1/seq-8
./sn95060791/1911/01/06/ed-1/seq-1
./sn95060791/1911/01/06/ed-1/seq-6
./sn95060791/1911/01/06/ed-1/seq-3
./sn95060791/1911/01/06/ed-1/seq-4
./sn95060791/1911/01/06/ed-1/seq-5
./sn95060791/1911/01/06/ed-1/seq-2
./sn95060791/1911/01/13
.

./sn95060791/1916/08
./sn95060791/1916/08/04
./sn95060791/1916/08/04/ed-1
./sn95060791/1916/08/04/ed-1/seq-1
./sn95060791/1916/08/04/ed-1/seq-3
./sn95060791/1916/08/04/ed-1/seq-4
./sn95060791/1916/08/04/ed-1/seq-2
./sn95060791/1916/08/18
./sn95060791/1916/08/18/ed-1
./sn95060791/1916/08/18/ed-1/seq-1
./sn95060791/1916/08/18/ed-1/seq-3
./sn95060791/1916/08/18/ed-1/seq-4
./sn95060791/1916/08/18/ed-1/seq-2
./sn95060791/1916/08/11
./sn95060791/1916/08/11/ed-1
./sn95060791/1916/08/11/ed-1/seq-1
./sn95060791/1916/08/11/ed-1/seq-3
./sn95060791/1916/08/11/ed-1/seq-4
./sn95060791/1916/08/11/ed-1/seq-2
./sn95060791/1916/08/25
./sn95060791/1916/08/25/ed-1
./sn95060791/1916/08/25/ed-1/seq-1
./sn95060791/1916/08/25/ed-1/seq-3
./sn95060791/1916/08/25/ed-1/seq-4
./sn95060791/1916/08/25/ed-1/seq-2
./sn95060791/1916/01
./sn95060791/1916/01/28
./sn95060791/1916/01/28/ed-1
./sn95060791/1916/01/28/ed-1/seq-1
./sn95060791/1916/01/28/ed-1/seq-6
./sn95060791/1916/01/28/ed-1/seq-3
./sn95060791/1916/01/28/ed-1

./sn95060791/1920/06/04
./sn95060791/1920/06/04/ed-1
./sn95060791/1920/06/04/ed-1/seq-1
./sn95060791/1920/06/04/ed-1/seq-3
./sn95060791/1920/06/04/ed-1/seq-4
./sn95060791/1920/06/04/ed-1/seq-2
./sn95060791/1920/06/18
./sn95060791/1920/06/18/ed-1
./sn95060791/1920/06/18/ed-1/seq-1
./sn95060791/1920/06/18/ed-1/seq-3
./sn95060791/1920/06/18/ed-1/seq-4
./sn95060791/1920/06/18/ed-1/seq-2
./sn95060791/1920/06/11
./sn95060791/1920/06/11/ed-1
./sn95060791/1920/06/11/ed-1/seq-1
./sn95060791/1920/06/11/ed-1/seq-3
./sn95060791/1920/06/11/ed-1/seq-4
./sn95060791/1920/06/11/ed-1/seq-2
./sn95060791/1920/06/25
./sn95060791/1920/06/25/ed-1
./sn95060791/1920/06/25/ed-1/seq-1
./sn95060791/1920/06/25/ed-1/seq-3
./sn95060791/1920/06/25/ed-1/seq-4
./sn95060791/1920/06/25/ed-1/seq-2
./sn95060791/1920/12
./sn95060791/1920/12/03
./sn95060791/1920/12/03/ed-1
./sn95060791/1920/12/03/ed-1/seq-1
./sn95060791/1920/12/03/ed-1/seq-3
./sn95060791/1920/12/03/ed-1/seq-4
./sn95060791/1920/12/03/ed-1/seq-2
./sn95060791/1

./sn95060791/1918/08/30/ed-1/seq-1
./sn95060791/1918/08/30/ed-1/seq-6
./sn95060791/1918/08/30/ed-1/seq-3
./sn95060791/1918/08/30/ed-1/seq-4
./sn95060791/1918/08/30/ed-1/seq-5
./sn95060791/1918/08/30/ed-1/seq-2
./sn95060791/1918/08/23
./sn95060791/1918/08/23/ed-1
./sn95060791/1918/08/23/ed-1/seq-1
./sn95060791/1918/08/23/ed-1/seq-6
./sn95060791/1918/08/23/ed-1/seq-3
./sn95060791/1918/08/23/ed-1/seq-4
./sn95060791/1918/08/23/ed-1/seq-5
./sn95060791/1918/08/23/ed-1/seq-2
./sn95060791/1918/01
./sn95060791/1918/01/04
./sn95060791/1918/01/04/ed-1
./sn95060791/1918/01/04/ed-1/seq-1
./sn95060791/1918/01/04/ed-1/seq-3
./sn95060791/1918/01/04/ed-1/seq-4
./sn95060791/1918/01/04/ed-1/seq-2
./sn95060791/1918/01/18
./sn95060791/1918/01/18/ed-1
./sn95060791/1918/01/18/ed-1/seq-1
./sn95060791/1918/01/18/ed-1/seq-3
./sn95060791/1918/01/18/ed-1/seq-4
./sn95060791/1918/01/18/ed-1/seq-2
./sn95060791/1918/01/11
./sn95060791/1918/01/11/ed-1
./sn95060791/1918/01/11/ed-1/seq-1
./sn95060791/1918/01/11/ed-1/seq

./sn95060791/1913/12
./sn95060791/1913/12/05
./sn95060791/1913/12/05/ed-1
./sn95060791/1913/12/05/ed-1/seq-1
./sn95060791/1913/12/05/ed-1/seq-3
./sn95060791/1913/12/05/ed-1/seq-4
./sn95060791/1913/12/05/ed-1/seq-2
./sn95060791/1913/12/19
./sn95060791/1913/12/19/ed-1
./sn95060791/1913/12/19/ed-1/seq-1
./sn95060791/1913/12/19/ed-1/seq-3
./sn95060791/1913/12/19/ed-1/seq-4
./sn95060791/1913/12/19/ed-1/seq-2
./sn95060791/1913/12/26
./sn95060791/1913/12/26/ed-1
./sn95060791/1913/12/26/ed-1/seq-1
./sn95060791/1913/12/26/ed-1/seq-3
./sn95060791/1913/12/26/ed-1/seq-4
./sn95060791/1913/12/26/ed-1/seq-2
./sn95060791/1913/12/12
./sn95060791/1913/12/12/ed-1
./sn95060791/1913/12/12/ed-1/seq-1
./sn95060791/1913/12/12/ed-1/seq-3
./sn95060791/1913/12/12/ed-1/seq-4
./sn95060791/1913/12/12/ed-1/seq-2
./sn95060791/1914
./sn95060791/1914/03
./sn95060791/1914/03/20
./sn95060791/1914/03/20/ed-1
./sn95060791/1914/03/20/ed-1/seq-1
./sn95060791/1914/03/20/ed-1/seq-3
./sn95060791/1914/03/20/ed-1/seq-4
./sn950607

./sn95060791/1915/03
./sn95060791/1915/03/05
./sn95060791/1915/03/05/ed-1
./sn95060791/1915/03/05/ed-1/seq-1
./sn95060791/1915/03/05/ed-1/seq-3
./sn95060791/1915/03/05/ed-1/seq-4
./sn95060791/1915/03/05/ed-1/seq-2
./sn95060791/1915/03/19
./sn95060791/1915/03/19/ed-1
./sn95060791/1915/03/19/ed-1/seq-1
./sn95060791/1915/03/19/ed-1/seq-3
./sn95060791/1915/03/19/ed-1/seq-4
./sn95060791/1915/03/19/ed-1/seq-2
./sn95060791/1915/03/26
./sn95060791/1915/03/26/ed-1
./sn95060791/1915/03/26/ed-1/seq-1
./sn95060791/1915/03/26/ed-1/seq-3
./sn95060791/1915/03/26/ed-1/seq-4
./sn95060791/1915/03/26/ed-1/seq-2
./sn95060791/1915/03/12
./sn95060791/1915/03/12/ed-1
./sn95060791/1915/03/12/ed-1/seq-1
./sn95060791/1915/03/12/ed-1/seq-3
./sn95060791/1915/03/12/ed-1/seq-4
./sn95060791/1915/03/12/ed-1/seq-2
./sn95060791/1915/04
./sn95060791/1915/04/02
./sn95060791/1915/04/02/ed-1
./sn95060791/1915/04/02/ed-1/seq-1
./sn95060791/1915/04/02/ed-1/seq-3
./sn95060791/1915/04/02/ed-1/seq-4
./sn95060791/1915/04/02/ed-1

./sn95060791/1912/03/29
./sn95060791/1912/03/29/ed-1
./sn95060791/1912/03/29/ed-1/seq-7
./sn95060791/1912/03/29/ed-1/seq-8
./sn95060791/1912/03/29/ed-1/seq-1
./sn95060791/1912/03/29/ed-1/seq-6
./sn95060791/1912/03/29/ed-1/seq-3
./sn95060791/1912/03/29/ed-1/seq-4
./sn95060791/1912/03/29/ed-1/seq-5
./sn95060791/1912/03/29/ed-1/seq-2
./sn95060791/1912/03/08
./sn95060791/1912/03/08/ed-1
./sn95060791/1912/03/08/ed-1/seq-7
./sn95060791/1912/03/08/ed-1/seq-8
./sn95060791/1912/03/08/ed-1/seq-1
./sn95060791/1912/03/08/ed-1/seq-6
./sn95060791/1912/03/08/ed-1/seq-3
./sn95060791/1912/03/08/ed-1/seq-4
./sn95060791/1912/03/08/ed-1/seq-5
./sn95060791/1912/03/08/ed-1/seq-2
./sn95060791/1912/03/01
./sn95060791/1912/03/01/ed-1
./sn95060791/1912/03/01/ed-1/seq-7
./sn95060791/1912/03/01/ed-1/seq-8
./sn95060791/1912/03/01/ed-1/seq-1
./sn95060791/1912/03/01/ed-1/seq-6
./sn95060791/1912/03/01/ed-1/seq-3
./sn95060791/1912/03/01/ed-1/seq-4
./sn95060791/1912/03/01/ed-1/seq-5
./sn95060791/1912/03/01/ed-1/seq-2
.

./sn95060791/1912/08/23/ed-1/seq-7
./sn95060791/1912/08/23/ed-1/seq-9
./sn95060791/1912/08/23/ed-1/seq-8
./sn95060791/1912/08/23/ed-1/seq-1
./sn95060791/1912/08/23/ed-1/seq-6
./sn95060791/1912/08/23/ed-1/seq-3
./sn95060791/1912/08/23/ed-1/seq-4
./sn95060791/1912/08/23/ed-1/seq-5
./sn95060791/1912/08/23/ed-1/seq-2
./sn95060791/1912/01
./sn95060791/1912/01/05
./sn95060791/1912/01/05/ed-1
./sn95060791/1912/01/05/ed-1/seq-7
./sn95060791/1912/01/05/ed-1/seq-8
./sn95060791/1912/01/05/ed-1/seq-1
./sn95060791/1912/01/05/ed-1/seq-6
./sn95060791/1912/01/05/ed-1/seq-3
./sn95060791/1912/01/05/ed-1/seq-4
./sn95060791/1912/01/05/ed-1/seq-5
./sn95060791/1912/01/05/ed-1/seq-2
./sn95060791/1912/01/19
./sn95060791/1912/01/19/ed-1
./sn95060791/1912/01/19/ed-1/seq-7
./sn95060791/1912/01/19/ed-1/seq-8
./sn95060791/1912/01/19/ed-1/seq-1
./sn95060791/1912/01/19/ed-1/seq-6
./sn95060791/1912/01/19/ed-1/seq-3
./sn95060791/1912/01/19/ed-1/seq-4
./sn95060791/1912/01/19/ed-1/seq-5
./sn95060791/1912/01/19/ed-1/seq-

./sn95060791/1909/12/31/ed-1/seq-5
./sn95060791/1909/12/31/ed-1/seq-2
./sn95060791/1909/12/24
./sn95060791/1909/12/24/ed-1
./sn95060791/1909/12/24/ed-1/seq-7
./sn95060791/1909/12/24/ed-1/seq-8
./sn95060791/1909/12/24/ed-1/seq-1
./sn95060791/1909/12/24/ed-1/seq-6
./sn95060791/1909/12/24/ed-1/seq-3
./sn95060791/1909/12/24/ed-1/seq-4
./sn95060791/1909/12/24/ed-1/seq-5
./sn95060791/1909/12/24/ed-1/seq-2


Now let's determine the number of directories (`-type d`) by piping the output of the `find` command to `wc` and counting the number of lines (`-l`) in the output.

In [51]:
!find . -type d | wc -l

   13306


(Notice that you only need to include the 'bang' at the beginning of the cell.)

My sample data contains multiple directories. We can inspect the first level of subdirectories using the `maxdepth` option.

In [52]:
!find . -type d -maxdepth 1

.
./sn89053851
./sn94051342
./sn94051341
./sn95060791


See more layers by increasing the -maxdepth value.

In [53]:
!find . -type d -maxdepth 2

.
./sn89053851
./sn89053851/1899
./sn89053851/1910
./sn89053851/1911
./sn89053851/1905
./sn89053851/1902
./sn89053851/1903
./sn89053851/1904
./sn89053851/1901
./sn89053851/1906
./sn89053851/1908
./sn89053851/1909
./sn89053851/1907
./sn89053851/1900
./sn94051342
./sn94051342/1910
./sn94051342/1911
./sn94051342/1913
./sn94051342/1912
./sn94051342/1901
./sn94051342/1908
./sn94051342/1909
./sn94051342/1907
./sn94051342/1900
./sn94051341
./sn94051341/1899
./sn94051341/1897
./sn94051341/1896
./sn94051341/1898
./sn94051341/1895
./sn94051341/1900
./sn95060791
./sn95060791/1919
./sn95060791/1921
./sn95060791/1917
./sn95060791/1910
./sn95060791/1911
./sn95060791/1916
./sn95060791/1920
./sn95060791/1918
./sn95060791/1913
./sn95060791/1914
./sn95060791/1915
./sn95060791/1912
./sn95060791/1909


Let's look at one of the subdirectories and count the directories in it. (Adding `./sn89053851` to the find command means start from the subdirectory named 'sn89053851'.)

In [54]:
!find ./sn89053851 -type d | wc -l

    3578


Now let's count the files (`-type f`), starting from the present working directory.

In [55]:
!find . -type f | wc -l

   18487


One more option: find files by size (`-size`). In this case, file of size greater than 1Mb (`+1M`).

In [56]:
!find . -type f -size +1M | wc -l

     604


Many, many tiny files slow down transfers. So let's use the command above to look for files smaller than 20Kb. (Note the use of a lowercase 'k', as opposed to an uppercase 'M' or 'G'.)

In [57]:
!find . -type f -size -20k | wc -l

    5685


For more information on the `find` and `wc` commands, use the `man` command. (While you don't need the '!' for this commnad, it does affect the display. Try running the cell with and without it.)

In [ ]:
!man find

In [ ]:
!man wc

### Looping one-liners

We can use looping (and other iterating techniques) to profile the files in a directory tree, recursively from the present working directory.

Here's a one-liner adapted from a command that John White used recently on Savio. It uses the `ls` command to list the subdirectories in the current directory, then lists the items in each subdirectory and passes the results to `wc` to count them.

In [58]:
!for i in `ls`; do echo $i; find $i | wc -l; done

sn89053851
    8287
sn94051341
    4783
sn94051342
    6256
sn95060791
   12465


This one-liner loops through a given set of values, calling the `find` command to identify files whose size is greater than each value, then pipes that information to `wc` to count those files.  Here, M = megabytes and G = gigabytes.

The output shows the number of files of each size.

In [59]:
!for size in +0 +1M +10M +1G +10G ; do echo "$size"; find . -type f -size $size | wc -l; done

+0
   18485
+1M
     604
+10M
       0
+1G
       0
+10G
       0


We can check our work with a little addition and one more command. The sum of the files and directories returned by John White's one-liner comes to 31,791. The number of files found by the one-liner just above is 18,486. (Why not 18,486 + 604? The 604 files greater than 1M in size are already counted in the number of files of size greater than 0.) The difference between the two -- i.e., the number of directories -- should be 13,305.

In [60]:
!find ./ -type d | wc -l

   13306


Off by one. That would be the containing directory (`batch_az_acacia_ver01`) not counted by the `ls` command in John's one-liner but noted by the `find` command that we just ran.

### Tree

The `tree` command diagrams the structure of your local file tree. The first argument points to the top of the tree you want to draw. The `-d` flag means 'show only directories' (i.e., not files); the `-o` option writes the output to a local file, which we'll name 'LOCTree.txt'. Writing to a file is valuable because the output can get quite long. 

I didn't have the 'tree' package installed on my Mac. I needed to use a package manager to download and install. I use Homebrew (https://brew.sh/), but there are others. To install 'tree' using Homebrew, the command is `brew install tree`. More documentation can be found at https://docs.brew.sh/Manpage.

In [61]:
!tree ./ -d -o LOCTree.txt

Find the output file just where you put it -- in the present working directory of the local file system.

In [62]:
ls

LOCTree.txt  sn89053851/  sn94051341/  sn94051342/  sn95060791/


Let's read the file into the notebook and take a look at the file tree.

In [65]:
cat "LOCTree.txt"

./
├── sn89053851
│   ├── 1899
│   │   ├── 06
│   │   │   └── 01
│   │   │       └── ed-1
│   │   │           ├── seq-1
│   │   │           ├── seq-10
│   │   │           ├── seq-11
│   │   │           ├── seq-12
│   │   │           ├── seq-2
│   │   │           ├── seq-3
│   │   │           ├── seq-4
│   │   │           ├── seq-5
│   │   │           ├── seq-6
│   │   │           ├── seq-7
│   │   │           ├── seq-8
│   │   │           └── seq-9
│   │   ├── 08
│   │   │   └── 31
│   │   │       └── ed-1
│   │   │           ├── seq-1
│   │   │           ├── seq-10
│   │   │           ├── seq-2
│   │   │           ├── seq-3
│   │   │           ├── seq-4
│   │   │           ├── seq-5
│   │   │           ├── seq-6
│   │   │           ├── seq-7
│   │   │           ├── seq-8
│   │   │           └── seq-9
│   │   ├── 09
│   │   │   ├── 07
│   │   │   │   └── ed-1
│   │   │   │       ├── seq-1
│   │   │   │       ├── seq-10
│   │   │   │       ├── seq-2
│   │   │   │       ├── seq-3
│   │  

│   │   │   │       ├── seq-2
│   │   │   │       ├── seq-3
│   │   │   │       └── seq-4
│   │   │   ├── 11
│   │   │   │   └── ed-1
│   │   │   │       ├── seq-1
│   │   │   │       ├── seq-2
│   │   │   │       ├── seq-3
│   │   │   │       └── seq-4
│   │   │   └── 18
│   │   │       └── ed-1
│   │   │           ├── seq-1
│   │   │           ├── seq-2
│   │   │           ├── seq-3
│   │   │           └── seq-4
│   │   ├── 08
│   │   │   ├── 01
│   │   │   │   └── ed-1
│   │   │   │       ├── seq-1
│   │   │   │       ├── seq-2
│   │   │   │       ├── seq-3
│   │   │   │       └── seq-4
│   │   │   ├── 08
│   │   │   │   └── ed-1
│   │   │   │       ├── seq-1
│   │   │   │       ├── seq-2
│   │   │   │       ├── seq-3
│   │   │   │       └── seq-4
│   │   │   ├── 15
│   │   │   │   └── ed-1
│   │   │   │       ├── seq-1
│   │   │   │       ├── seq-2
│   │   │   │       ├── seq-3
│   │   │   │       └── seq-4
│   │   │   ├── 22
│   │   │   │   └── ed-1
│   │   │   │       ├── seq-1
│

    │   │   ├── 12
    │   │   │   └── ed-1
    │   │   │       ├── seq-1
    │   │   │       ├── seq-2
    │   │   │       ├── seq-3
    │   │   │       └── seq-4
    │   │   ├── 19
    │   │   │   └── ed-1
    │   │   │       ├── seq-1
    │   │   │       ├── seq-2
    │   │   │       ├── seq-3
    │   │   │       ├── seq-4
    │   │   │       └── seq-5
    │   │   └── 26
    │   │       └── ed-1
    │   │           ├── seq-1
    │   │           ├── seq-2
    │   │           ├── seq-3
    │   │           └── seq-4
    │   └── 12
    │       ├── 03
    │       │   └── ed-1
    │       │       ├── seq-1

    │       │       ├── seq-3
    │       │       └── seq-4
    │       ├── 10
    │       │   └── ed-1
    │       │       ├── seq-1
    │       │       ├── seq-2
    │       │       ├── seq-3
    │       │       ├── seq-4
    │       │       ├── seq-5
    │       │       ├── seq-6
    │       │       ├── seq-7
    │       │       └── seq-8
    │       ├── 17
    │       │   └── ed-1


    │   │   │       ├─�� seq-3
    │   │   │       └── seq-4
    │   │   ├── 15
    │   │   │   └── ed-1
    │   │   │       ├── seq-1
    │   │   │       ├── seq-2
    │   │   │       ├── seq-3
    │   │   │       └── seq-4
    │   │   └── 22
    │   │       └── ed-1
    │   │           ├── seq-1
    │   │           ├── seq-2
    │   │           ├── seq-3
    │   │           └── seq-4
    │   ├── 03
    │   │   ├── 01
    │   │   │   └── ed-1
    │   │   │       ├── seq-1
    │   │   │       ├── seq-2
    │   │   │       ├── seq-3
    │   │   │       └── seq-4
    │   │   ├── 08
    │   │   │   └── ed-1
    │   │   │       ├── seq-1
    │   │   │       ├── seq-2
    │   │   │       ├── seq-3
    │   │   │       └── seq-4
    │   │   ├── 15
    │   │   │   └── ed-1
    │   │   │       ├── seq-1
    │   │   │       ├── seq-2
    │   │   │       ├── seq-3
    │   │   │       └── seq-4
    │   │   ├── 22
    │   │   │   └── ed-1
    │   │   │       ├── seq-1
    │   │   │       ├── seq-2


FYI, you could also open the file in the notebook using Python, and read it:

In [ ]:
with open("LOCTree.txt", "r") as f:
    print(f.read())

Tree includes an option to limit how many levels of subdirectories to assess. That's useful for getting a sense of the structure of your file tree.

In [66]:
!tree ./ -d -L 1

./
├── sn89053851
├── sn94051341
├── sn94051342
└── sn95060791

4 directories


In [67]:
!tree ./ -d -L 2

./
├── sn89053851
│   ├── 1899
│   ├── 1900
│   ├── 1901
│   ├── 1902
│   ├── 1903
│   ├── 1904
│   ├── 1905
│   ├── 1906
│   ├── 1907
│   ├── 1908
│   ├── 1909
│   ├── 1910
│   └── 1911
├── sn94051341
│   ├── 1895
│   ├── 1896
│   ├── 1897
│   ├── 1898
│   ├── 1899
│   └── 1900
├── sn94051342
│   ├── 1900
│   ├── 1901
│   ├── 1907
│   ├── 1908
│   ├── 1909
│   ├── 1910
│   ├── 1911
│   ├── 1912
│   └── 1913
└── sn95060791
    ├── 1909
    ├── 1910
    ├── 1911
    ├── 1912
    ├── 1913
    ├── 1914
    ├── 1915
    ├── 1916
    ├── 1917
    ├── 1918
    ├── 1919
    ├── 1920
    └── 1921

45 directories


In [68]:
!tree ./ -d -L 3

./
├── sn89053851
│   ├── 1899
│   │   ├── 06
│   │   ├── 08
│   │   ├── 09
│   │   ├── 10
│   │   ├── 11
│   │   └── 12
│   ├── 1900
│   │   ├── 01
│   │   ├── 03
│   │   ├── 04
│   │   ├── 05
│   │   ├── 06
│   │   ├── 07
│   │   ├── 08
│   │   ├── 09
│   │   ├── 10
│   │   ├── 11
│   │   └── 12
│   ├── 1901
│   │   ├── 01
│   │   ├── 02
│   │   ├── 03
│   │   ├── 04
│   │   ├── 05
│   │   ├── 06
│   │   ├── 07
│   │   ├── 08
│   │   ├── 09
│   │   ├── 10
│   │   ├── 11
│   │   └── 12
│   ├── 1902
│   │   ├── 01
│   │   ├── 02
│   │   ├── 03
│   │   ├── 04
│   │   ├── 05
│   │   ├── 06
│   │   ├── 07
│   │   ├── 08
│   │   ├── 09
│   │   ├── 10
│   │   ├── 11
│   │   └── 12
│   ├── 1903
│   │   ├── 01
│   │   ├── 02
│   │   ├── 03
│   │   ├── 04
│   │   ├── 05
│   │   ├── 06
│   │   ├── 07
│   │   ├── 08
│   │   ├── 09
│   │   ├── 10
│   │   ├── 11
│   │   └── 12
│   ├── 1904
│   │   ├── 01
│   │   ├── 02
│   │   ├── 03
│   │   ├── 04
│   │   ├── 05
│   │   ├── 06
│   │   ├── 07
│   

We could have done the previous three commands as one, iterating over a range of numbers, but the output becomes a little confusing.

In [ ]:
!for i in 1 2 3; do tree ./ -d -L $i; done

##  Running shell scripts

Can we run a shell script? Yes. (Of course.)

For instructions, see: https://www.quora.com/How-do-I-execute-bash-scripts-via-IPython-Jupyter-notebook.

Let's demonstrate how-to by running a useful script...one that finds all directories with greater than 2000 files and displays the size and name of those directories. 
 
 First, let's locate the script: it's called 'dirs-gt-2000-du.sh' and it's in the same directory as this notebook. See In[2], above.

In [69]:
cd /Users/rjaffe/Jupyter_notebooks/rdm_datatransfer

/Users/rjaffe/Jupyter_notebooks/rdm_datatransfer


In [70]:
ls

Shell_play.ipynb    dirs-gt-2000-du.sh* dirs-gt-2000.sh*    readme.md


To be sure that the script does what we expect, let's look at it.

In [71]:
cat "dirs-gt-2000-du.sh"

#!/bin/bash
topdir=$1
for dir in `find $topdir -type d` ; do
  count=`find $dir -type f | wc -l`
  if [ $count -gt 2000 ]
  then
    size=`du -d=0 -h $dir`
	echo $size
  fi
done

Running scripts within the notebook requires starting a new shell process. There's a library for that: `subprocess`.

In [72]:
import subprocess

The script takes an argument: the directory that we will profile. We pass both the script name and directory to be profiled as arguments to the subprocess function. 

Be patient: this will take a while!

In [73]:
subprocess.run(['./dirs-gt-2000-du.sh', '/Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01'])

CompletedProcess(args=['./dirs-gt-2000-du.sh', '/Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01'], returncode=0)

Hmmmm...something's wrong. On my local system, the output reads:

    5.7G /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01
    1.9G /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/sn89053851
    898M /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/sn94051342
    913M /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/sn94051341
    2.1G /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/sn95060791

Here's what's up: `subprocess.run` returns a CompletedProcess instance that shows the arguments we provided and the returncode of `0` produced by the `echo` command. That's what gets displayed in the cell's output. To see the *output* of the echo command, we must ignore the return code (i.e., send it to `dev/null`) and `pipe` the process's output to `stdout`, per: 
https://docs.python.org/3/library/subprocess.html#subprocess.RUN. 

As the documentation notes, "By default, this function returns the data as encoded bytes." We decode it to UTF-8.

In [74]:
dirinfo = subprocess.run(['./dirs-gt-2000-du.sh', '/Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01', 'null'], stdout=subprocess.PIPE)
print(dirinfo.stdout.decode("utf-8", "strict"))

5.7G /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01
1.9G /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/sn89053851
898M /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/sn94051342
913M /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/sn94051341
2.1G /Users/rjaffe/Documents/RDM/TestData/LOC_Data/batch_az_acacia_ver01/sn95060791



voilà!

## Challenge

PROBLEM: RDM has consulted on several cases in which researchers were trying to upload millions of files to Box or Drive and encountering v-e-r-y slow transfer rates. Also, we've encountered at least two common research tools (Opera Phenix microscopes and FMRIB FSL analysis software) that can produce individual directories containing greater than 20,000 files -- the limit that Box can display without throwing a "File can't be found' error in response to the MLSD command used by FileZilla and other FTP clients.

CONSTRAINTS:

• Copying zillions of tiny files ("ZOTS") takes much longer than many fewer, larger files. 

• Google Drive and Box both throttle bulk uploads to as few as two files per second. 

• Box has a per-file file size limit of 15GB.  

CHALLENGE 1: Can we use these tools to devise a strategy for identifying the ends of the tree -- where the number of files tend to be great -- and then tarring or zipping up just those directories? The goal is to enable faster transfer rates to Box and Drive, while allowing researchers to download to their analysis environments only those portions of the data that they need for any given analysis.

CHALLENGE 2: Can we run this notebook from the cloud and point it at a local machine with internal or mounted storage to transfer?

# Using rclone

Now let's try running rclone to copy a file from the local machine to Google Drive.

First, navigate to the file...

In [ ]:
cd /Users/rjaffe/Documents/RDM/TestData

In [ ]:
ls

In [ ]:
cd Transfer-test-no-subfolders/

In [ ]:
ls -al

Now run rclone to copy one of the files in this directory to a pre-configured drive account.

In [ ]:
!rclone copy /Users/rjaffe/Documents/RDM/TestData/Transfer-test-no-subfolders/Picture\ 8.png rclone-rdmconsult-bdrive:test_from_notebook

Check the drive account in a separate browser to confirm that the file actually transferred. (It did!)

# Using Globus and the Globus API

...to be added.